presto-main/src/main/java/io/prestosql/execution/scheduler/SourcePartitionedScheduler.java

```diff
@@ -478,6 +501,16 @@ public void close()
         return newTasks.build();
     }
 
+    private Set<RemoteTask> createTaskOnRandomNode()
+    {
+        List<InternalNode> allNodes = splitPlacementPolicy.allNodes();
+        checkState(allNodes.size() > 0, "No nodes available");
+        int position = ThreadLocalRandom.current().nextInt(0, allNodes.size());
+        InternalNode node = allNodes.get(position);
+        checkState(!stage.getScheduledNodes().contains(node), "Stage task is already scheduled on node");
```

Comment:
Should `stage.getScheduledNodes()` be empty?  If so, I'd check it at the start of the method.